Fixes to enable libktx use of Basis

MarkCallow opened the pull request:
This PR does the following:

Enables access to the low-level decoder:
- Makes a new signature for transcode_slice that does not require faking up bits of a .basis file. A function with the old signature remains as an inline function that calls the new one.
- Makes `write_opaque_blocks` a public static function of `lowlevel_decoder.`

Enables embedding of the basis code within a library to be linked to applications:
- Changes the setting of _ITERATOR_DEBUG_LEVEL to the default VS value. There appears to be absolutely no need to set this value or _SECURE_SCL within these files.

Fixes all the unused variable warnings which totalled about 45.

It does NOT fix other -pedantic warnings which include one about misuse of a comma operator and 9 about anonymous structs and unions being extensions. There is also 1 unused variable warning that appears when compiled in release mode that this PR does not fix.

Please turn on -pedantic and fix them.

donmccurdy commented:
> Makes a new signature for transcode_slice that does not require faking up bits of a .basis file.

I think this new signature, and access to the low-level decoder, will also be helpful for the WASM transcoder. In a glTF loader, for example, we would have metadata available and want to transcode the texture without constructing a `.basis` container at runtime.

jherico commented:
@MarkCallow if you want to make a version of this PR against my fork, which is used for the vcpkg version of basis, I'll merge it.

atteneder commented:
I [fixed](https://github.com/atteneder/basis_universal/tree/ktxfixes) the current conflicts and filed a sub PR:

https://github.com/MarkCallow/basis_universal/pull/1

MarkCallow commented:
I've updated this to be against the current tip of master which means the new function signature for transcode_slice has been updated to accept in the extra arguments added by the recent transcoder update and the in-line implementation of the original function has been similarly updated.

In doing so I removed 2 changes that are no longer necessary:
1. the commenting out of an assert in basisu_enc.h. The problem that cause the assert to fire appears to have been fixed.
2. the override of the _ITERATOR_DEBUG_LEVEL setting. BasisU now provides a ifdef to override the setting.

This PR now fixes all the warnings found by gcc and clang, except those involving anonymous structs and unions. I could no longer cope with the vast number of warnings masking real problems in my own code. These included unused tables as well as variables. Most of these have been commented or ifdef'ed out. A few have been wrapped in a new DE_MAYBE_UNUSED macro as they are used when compiled in debug mode. Other warnings involved suspicious use of a comma operator, data truncation and attempts to use multi-character character constants.

In my own build environment I have suppressed the warnings for anonymous structs and unions as all the compilers I am using actually support them (they're just not valid in C++11) and the code would become significantly poorer if their use was avoided.

MarkCallow commented:
@jherico `master` branch of your fork appears to differ significantly from `master` branch of this repo. E.g., most of the transcoder files are missing. Since the most important part of this PR is changes to the missing transcoder files there seems no point in making a PR for you.

jherico commented:
@MarkCallow my master branch is based off the changes in #13 so the files are in different locations.  I suppose I could merge from upstream, but honestly there hasn't been much in the way of substantial changes.

MarkCallow commented:
> but honestly there hasn't been much in the way of substantial changes.

There is one substantial change you appear to be missing. It looks like you do not have all the latest transcoders or at least not in a way where you are building them. The transcoder tables for ASTC, bc7_m5 and atc are all in the transcoders folder while the other transcode files are in src/transcode. PVRTC2 support seems to be completely missing.

jherico commented:
@MarkCallow Sorry, I guess I missed some significant work that happened shortly after my last upstream merge.  I've merged from master again this morning.

The `inc` files are still split between the `/transcoder` and `/src/transcoder` folders, but I'm going to leave that be for the moment as it doesn't seem to be negatively impacting the build.

jherico commented:
@MarkCallow I've also merged this PR into a ktxfixes branch in my own repo, so you don't need a separate PR against my fork.

MarkCallow commented:
> I've also merged this PR into a ktxfixes branch in my own repo

Cool. Thanks.

richgel999 commented:
Note the transcoder in Basis Universal is basically done, apart from bug fixes or more optimizations. So you shouldn't have to merge any more.

MarkCallow commented:
Obsoleted by #146.
